File: backend/app/models/tool_catalog.py
from sqlalchemy import BigInteger, Boolean, String, Text, select
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, Session, mapped_column

from app.core.db import Base
from app.models.mixins import TimestampMixin


class ToolCatalog(TimestampMixin, Base):
    """도구 카탈로그 — 우리가 시드로 넣는 지원 도구(MCP·API) 목록. 참조/lookup 테이블.

    - type: "mcp" | "api"
    - auth_owner: "user"(유저 본인 붙여넣기) | "developer"(우리 공용키)
    - metadata_json: 키 붙여넣기 팝업 세부(field·help·placeholder 등)
    """

    __tablename__ = "tool_catalog"

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True)
    key: Mapped[str] = mapped_column(String(100), unique=True, nullable=False, index=True)
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    key_required: Mapped[bool] = mapped_column(Boolean, nullable=False)
    key_issue_url: Mapped[str | None] = mapped_column(String(255), nullable=True)
    metadata_json: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    type: Mapped[str] = mapped_column(String(20), nullable=False)
    auth_owner: Mapped[str] = mapped_column(String(20), nullable=False)


def list_catalog_keys(db: Session) -> list[str]:
    """전체 카탈로그 key 목록 (assemble/skills/workflows 라우터가 공유하는 단일 소스).

    "실제 존재하는 도구인가"를 검증하는 모든 곳(AI 프롬프트 제약, 발행 시 used_mcps
    필터링, 목록 조회 시 스푸핑 방지)이 이 하나의 쿼리 결과를 기준으로 판단한다.
    """
    return list(db.scalars(select(ToolCatalog.key).order_by(ToolCatalog.key)))

File: frontend/src/lib/toolCatalog.ts
// 도구 카탈로그(GET /tool-catalog) 응답 타입. 키 붙여넣기 팝업을 이 메타로 자동 생성한다.

/** 입력칸 하나. name은 실행기가 MCP 서버에 넘길 환경변수명과 일치해야 한다
 *  (local-runner/app/core/mcp.py 의 MCP_SERVERS[*].env_fields). */
export interface ToolMetaField {
  name: string;
  placeholder?: string;
  help?: string;
}

export interface ToolMeta {
  /** 입력칸 여러 개 (예: 텔레그램=토큰+chat_id). 값은 JSON으로 묶여 저장된다. */
  fields?: ToolMetaField[];
  /** 발급 절차 안내. 링크 하나만 던지면 사용자가 헤매서, 단계로 보여준다. */
  guide?: string[];
  /** @deprecated fields를 쓸 것. 입력칸이 하나뿐이던 시절의 형식 */
  field?: string;
  /** @deprecated fields[].help 를 쓸 것 */
  help?: string;
  /** @deprecated fields[].placeholder 를 쓸 것 */
  placeholder?: string;
}

export interface ToolCatalogItem {
  id: number;
  key: string;
  name: string;
  type: string; // mcp / api
  auth_owner: string; // user(본인 붙여넣기) / developer(공용키)
  key_required: boolean;
  key_issue_url: string | null;
  description: string | null;
  metadata_json: ToolMeta | null;
}
